url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/sts/nusmv/NuSMVConsole.java

```diff
@@ -0,0 +1,289 @@
+package eu.fbk.soa.evolution.sts.nusmv;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.log4j.Logger;
+
+import eu.fbk.soa.evolution.sts.impl.NusmvSTS;
+import eu.fbk.soa.evolution.sts.impl.NusmvState;
+import eu.fbk.soa.util.IOUtils;
+
+
+class NuSMVConsole extends Thread {
+	
+	static int simulationSteps = 60;
+	
+	static Logger logger = Logger.getLogger(NuSMVConsole.class);
+	
+	private static String cmd_prompt = "NuSMV > ";
+	
+	private String nusmvFile;
+	
+	private String nusmvPath;
+	
+//	private String outFilePath;
+	
+	private File outFile;
+
+	private NusmvSTS sts;
+	
+	private Process proc;
+	
+	private BufferedReader brInput;
+	
+	private BufferedWriter bwOutput;
+
+	private StringBuffer sbTotal;
+	
+	private String consolePrint;
+    
+	private String strTotal = "";
+	
+	private Node initialNode;
+	
+	private Node currentNode;
+	
+//	private sts.State currentState;	
+
+	NuSMVConsole(String nusmvFile, String nusmvPath, String outFilePath, NusmvSTS sts) {
+		this.nusmvFile = nusmvFile;		
+		this.sbTotal = new StringBuffer();		
+		this.nusmvPath = nusmvPath;
+		this.outFile = new File(outFilePath);
+		this.sts = sts;
+	}
+
+	void readChars() {
+		try {
+			if (brInput.ready()) {
+				char[] cbuf = new char[2048];
+				int nrOfChars = brInput.read(cbuf, 0, 2048);
+				if (nrOfChars > 0) {
+					int i = 0;
+					while (i < nrOfChars) {
+						sbTotal.append(cbuf[i]);
+						strTotal += cbuf[i];
+						i++;
+					}
+				}
+			} 
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+
+	private synchronized String readStr() {
+		return strTotal;
+	}
+
+	private synchronized String printConsole() {
+		consolePrint = sbTotal.toString();
+		sbTotal = new StringBuffer();
+		strTotal = "";
+		return consolePrint;
+	}
+
+	void closeConsole() {
+		proc.destroy();
+	}
+
+	String waitForCommandPrompt() throws IOException {
+		while(! readStr().endsWith(cmd_prompt)) {
+			try {
+				Thread.sleep(50);				
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+		}
+		synchronized(this) {
+			return printConsole();
+		}
+	}
+
+	
+	
+	String pickInitialState(StateTree pathsToExplore) throws IOException {
+		
+		execCommand("bmc_pick_state -v -i");
+		try {
+			Thread.sleep(80);
+		} catch (InterruptedException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+		String fileContents = IOUtils.readFileAsString(outFile);
+		List<eu.fbk.soa.evolution.sts.State> initialStates = 
+			NuSMVOutputHandler.addInitialStatesToSTS(fileContents, sts);
+		
+		while (initialStates.isEmpty()) {
+			try {
+				logger.info("Read file to soon, waiting..");
+				Thread.sleep(200);
+			} catch (InterruptedException ie) {
+				ie.printStackTrace();
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 129:134, 231:236.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.